crates/autopilot/src/solvable_orders.rs

```diff
@@ -658,14 +658,35 @@ fn to_normalized_price(price: f64) -> Option<U256> {
 
 async fn filter_unsupported_tokens(
     mut orders: Vec<Order>,
-    bad_token: &dyn BadTokenDetecting,
+    bad_token: Arc<dyn BadTokenDetecting>,
 ) -> Result<Vec<Order>> {
-    // Can't use normal `retain` or `filter` because the bad token detection is
-    // async. So either this manual iteration or conversion to stream.
+    let bad_tokens = join_all(
+        orders
+            .iter()
+            .flat_map(|o| o.data.token_pair().unwrap_or_default())
+            .unique()
+            .map(|token| {
+                let bad_token = bad_token.clone();
+                async move {
+                    match bad_token.detect(token).await {
+                        Ok(quality) => (!quality.is_good()).then_some(token),
+                        Err(err) => {
+                            tracing::warn!(?token, ?err, "unable to determine token quality");
+                            Some(token)
+                        }
+                    }
+                }
+            }),
+    )
+    .await
+    .into_iter()
+    .flatten()
+    .collect::<HashSet<_>>();
+
     let mut index = 0;
     'outer: while index < orders.len() {
```

Comment:
This weird loop was just to work around the `async` calls. Now we can simply use `.retain()` to make this a lot more readable.